File: backend/app/services/churn_service.js
import { PreprocessingService } from './preprocessing_service.js';
import { defaultConfig } from '../config/config.js';

/**
 * XGBoost Tree Node for Gradient Boosted Decision Trees
 */
export class XGBoostTree {
  constructor(maxDepth = 4, minSamplesSplit = 3, l2Reg = 1.0) {
    this.maxDepth = maxDepth;
    this.minSamplesSplit = minSamplesSplit;
    this.l2Reg = l2Reg; // Lambda regularization
    this.root = null;
  }

  train(X, g, h, depth = 0, featureSubsetRatio = 0.8) {
    const numSamples = X.length;
    const numFeatures = X[0] ? X[0].length : 0;

    const sumG = g.reduce((a, b) => a + b, 0);
    const sumH = h.reduce((a, b) => a + b, 0);

    // Leaf weight calculation according to XGBoost objective: w = sum(g) / (sum(h) + lambda)
    const leafWeight = sumG / (sumH + this.l2Reg);

    // Stopping criteria
    if (depth >= this.maxDepth || numSamples < this.minSamplesSplit || sumH === 0) {
      return { isLeaf: true, weight: leafWeight };
    }

    // Subsample features (colsample_bytree)
    const featureIndices = [];
    const numToSelect = Math.max(1, Math.floor(numFeatures * featureSubsetRatio));
    while (featureIndices.length < numToSelect) {
      const idx = Math.floor(Math.random() * numFeatures);
      if (!featureIndices.includes(idx)) featureIndices.push(idx);
    }

    let bestGain = -1e9;
    let bestFeatureIdx = -1;
    let bestThreshold = 0;
    let bestSplits = null;

    const parentScore = (sumG * sumG) / (sumH + this.l2Reg);

    featureIndices.forEach(fIdx => {
      // Collect unique feature values for threshold candidates
      const values = X.map(row => row[fIdx]).sort((a, b) => a - b);
      const thresholds = [];
      const step = Math.max(1, Math.floor(values.length / 10));
      for (let i = 0; i < values.length - 1; i += step) {
        thresholds.push((values[i] + values[i + 1]) / 2);
      }

      thresholds.forEach(thresh => {
        let leftG = 0, leftH = 0;
        let rightG = 0, rightH = 0;
        const leftX = [], rightX = [];
        const leftGArr = [], rightGArr = [];
        const leftHArr = [], rightHArr = [];

        for (let i = 0; i < numSamples; i++) {
          if (X[i][fIdx] <= thresh) {
            leftX.push(X[i]);
            leftGArr.push(g[i]);
            leftHArr.push(h[i]);
            leftG += g[i];
            leftH += h[i];
          } else {
            rightX.push(X[i]);
            rightGArr.push(g[i]);
            rightHArr.push(h[i]);
            rightG += g[i];
            rightH += h[i];
          }
        }

        if (leftX.length === 0 || rightX.length === 0) return;

        const leftScore = (leftG * leftG) / (leftH + this.l2Reg);
        const rightScore = (rightG * rightG) / (rightH + this.l2Reg);

        // XGBoost Gain formula: Gain = 0.5 * (LeftScore + RightScore - ParentScore)
        const gain = 0.5 * (leftScore + rightScore - parentScore);

        if (gain > bestGain) {
          bestGain = gain;
          bestFeatureIdx = fIdx;
          bestThreshold = thresh;
          bestSplits = {
            leftX, leftGArr, leftHArr,
            rightX, rightGArr, rightHArr
          };
        }
      });
    });

    if (bestGain <= 0 || !bestSplits) {
      return { isLeaf: true, weight: leafWeight };
    }

    const leftChild = this.train(
      bestSplits.leftX, bestSplits.leftGArr, bestSplits.leftHArr,
      depth + 1, featureSubsetRatio
    );
    const rightChild = this.train(
      bestSplits.rightX, bestSplits.rightGArr, bestSplits.rightHArr,
      depth + 1, featureSubsetRatio
    );

    this.root = {
      isLeaf: false,
      featureIndex: bestFeatureIdx,
      threshold: bestThreshold,
      left: leftChild,
      right: rightChild,
    };

    return this.root;
  }

  predictWeight(row, node = this.root) {
    if (!node) return 0;
    if (node.isLeaf) return node.weight || 0;

    if (node.featureIndex !== undefined && node.threshold !== undefined) {
      if (row[node.featureIndex] <= node.threshold) {
        return this.predictWeight(row, node.left || null);
      } else {
        return this.predictWeight(row, node.right || null);
      }
    }
    return 0;
  }
}

/**
 * XGBoost Churn Prediction Service
 */
export class ChurnService {
  constructor() {
    this.trees = [];
    this.baseMargin = 0; // Initial log-odds margin F0
    this.learningRate = defaultConfig.modelParams.learningRate || 0.1;
    this.processedData = null;
    this.schema = null;
    this.metrics = null;
    this.globalFeatureImportance = [];
    this.thresholds = defaultConfig.thresholds;
  }

  setThresholds(newThresholds) {
    this.thresholds = newThresholds;
  }

  getThresholds() {
    return this.thresholds;
  }

  sigmoid(x) {
    return 1 / (1 + Math.exp(-Math.max(-15, Math.min(15, x))));
  }

  logOdds(p) {
    const clampedP = Math.max(0.001, Math.min(0.999, p));
    return Math.log(clampedP / (1 - clampedP));
  }

  trainModel(members, schema) {
    this.schema = schema;
    this.processedData = PreprocessingService.processDataset(
      members,
      schema,
      defaultConfig.modelParams.trainRatio
    );

    const { XTrain, yTrain, XTest, yTest, featureNames, featureLabels } = this.processedData;

    // 1. Initial base margin calculation (log-odds of target prior)
    const positiveCount = yTrain.filter(y => y === 1).length;
    const priorP = positiveCount / (yTrain.length || 1);
    this.baseMargin = this.logOdds(priorP);

    const nSamples = XTrain.length;
    const rawMargins = new Array(nSamples).fill(this.baseMargin);

    const nEstimators = defaultConfig.modelParams.nEstimators || 40;
    const maxDepth = defaultConfig.modelParams.maxDepth || 4;
    const minSamplesSplit = defaultConfig.modelParams.minSamplesSplit || 3;
    const l2Reg = defaultConfig.modelParams.l2Reg || 1.0;
    this.learningRate = defaultConfig.modelParams.learningRate || 0.1;

    this.trees = [];

    // 2. Iterative XGBoost Gradient Boosting Loop
    for (let m = 0; m < nEstimators; m++) {
      // Compute gradients (g) and hessians (h) for binary log-loss
      const g = [];
      const h = [];

      for (let i = 0; i < nSamples; i++) {
        const p = this.sigmoid(rawMargins[i]);
        // Gradient: g = y - p (residual)
        g.push(yTrain[i] - p);
        // Hessian: h = p * (1 - p)
        h.push(p * (1 - p));
      }

      // Train tree to predict gradients
      const tree = new XGBoostTree(maxDepth, minSamplesSplit, l2Reg);
      tree.train(XTrain, g, h);
      this.trees.push(tree);

      // Update raw margins for training samples
      for (let i = 0; i < nSamples; i++) {
        const treePred = tree.predictWeight(XTrain[i]);
        rawMargins[i] += this.learningRate * treePred;
      }
    }

    // 3. Evaluate on Test Set
    let tp = 0, fp = 0, tn = 0, fn = 0;
    const testProbs = [];

    XTest.forEach((row, i) => {
      const prob = this.predictProbFromRow(row);
      testProbs.push(prob);

      const actual = yTest[i];
      const predicted = prob >= 0.5 ? 1 : 0;

      if (actual === 1 && predicted === 1) tp++;
      else if (actual === 0 && predicted === 1) fp++;
      else if (actual === 0 && predicted === 0) tn++;
      else if (actual === 1 && predicted === 0) fn++;
    });

    const totalTest = XTest.length || 1;
    const accuracy = (tp + tn) / totalTest;
    const precision = (tp + fp) > 0 ? tp / (tp + fp) : 0;
    const recall = (tp + fn) > 0 ? tp / (tp + fn) : 0;
    const f1Score = (precision + recall) > 0 ? (2 * precision * recall) / (precision + recall) : 0;

    const rocAuc = this.calculateRocAuc(testProbs, yTest);

    this.metrics = {
      modelType: 'XGBoost (Gradient Boosted Trees)',
      accuracy: parseFloat(accuracy.toFixed(4)),
      precision: parseFloat(precision.toFixed(4)),
      recall: parseFloat(recall.toFixed(4)),
      f1Score: parseFloat(f1Score.toFixed(4)),
      rocAuc: parseFloat(rocAuc.toFixed(4)),
      confusionMatrix: { truePositive: tp, falsePositive: fp, trueNegative: tn, falseNegative: fn },
      trainSize: XTrain.length,
      testSize: XTest.length,
    };

    // 4. Compute Global Feature Importance
    this.calculateGlobalImportance();

    return { metrics: this.metrics, featureImportance: this.globalFeatureImportance };
  }

  predictProbFromRow(row) {
    let margin = this.baseMargin;
    for (let i = 0; i < this.trees.length; i++) {
      margin += this.learningRate * this.trees[i].predictWeight(row);
    }
    const prob = this.sigmoid(margin);
    return parseFloat(prob.toFixed(4));
  }

  predictMember(member) {
    if (!this.processedData || !this.schema) {
      throw new Error('XGBoost model not trained yet.');
    }

    const row = PreprocessingService.transformSingleMember(member, this.schema, this.processedData);
    const prob = this.predictProbFromRow(row);
    const riskLevel = this.classifyRiskLevel(prob);
    const prediction = prob >= 0.5 ? 'Churn' : 'Retained';

    return { probability: prob, riskLevel, prediction };
  }

  classifyRiskLevel(prob) {
    if (prob < this.thresholds.lowMax) return 'LOW';
    if (prob <= this.thresholds.mediumMax) return 'MEDIUM';
    return 'HIGH';
  }

  getMetrics() {
    if (!this.metrics) {
      throw new Error('XGBoost model metrics not available.');
    }
    return this.metrics;
  }

  getGlobalFeatureImportance() {
    return this.globalFeatureImportance;
  }

  getProcessedDataset() {
    return this.processedData;
  }

  getSchema() {
    return this.schema;
  }

  calculateRocAuc(probs, actuals) {
    const positives = probs.filter((_, i) => actuals[i] === 1);
    const negatives = probs.filter((_, i) => actuals[i] === 0);

    if (positives.length === 0 || negatives.length === 0) return 0.85;

    let wins = 0;
    positives.forEach(p => {
      negatives.forEach(n => {
        if (p > n) wins += 1.0;
        else if (p === n) wins += 0.5;
      });
    });

    const auc = wins / (positives.length * negatives.length);
    return Math.max(0.5, Math.min(1.0, auc));
  }

  calculateGlobalImportance() {
    if (!this.processedData) return;

    const { featureNames, featureLabels, X } = this.processedData;
    const numFeatures = featureNames.length;
    const importanceScores = new Array(numFeatures).fill(0);

    // Baseline predictions
    const baseProbs = X.map(row => this.predictProbFromRow(row));

    // Permutation sensitivity for each feature under XGBoost
    featureNames.forEach((_, fIdx) => {
      let diffSum = 0;
      X.forEach((row, rIdx) => {
        const rowCopy = [...row];
        rowCopy[fIdx] = rowCopy[fIdx] === 0 ? 1 : rowCopy[fIdx] * 1.5 + 1;
        const newProb = this.predictProbFromRow(rowCopy);
        diffSum += Math.abs(newProb - baseProbs[rIdx]);
      });
      importanceScores[fIdx] = diffSum;
    });

    const totalScore = importanceScores.reduce((a, b) => a + b, 0) || 1;

    const list = featureNames.map((name, i) => ({
      feature: name,
      featureLabel: featureLabels[name] || name,
      importance: parseFloat((importanceScores[i] / totalScore).toFixed(4)),
    }));

    list.sort((a, b) => b.importance - a.importance);
    this.globalFeatureImportance = list;
  }
}

File: backend/app/services/churn_service.py
import os
import joblib
import numpy as np
import pandas as pd
import catboost as cb
from pathlib import Path
from typing import Dict, Any, List, Tuple, Optional
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    confusion_matrix,
)

from app.config.settings import settings
from app.services.preprocessing_service import PreprocessingService
from app.utils.helpers import safe_float

class ChurnService:
    def __init__(self):
        self.model: Optional[cb.CatBoostClassifier] = None
        self.optimal_threshold: float = 0.5
        self.processed_data: Optional[Dict[str, Any]] = None
        self.schema: Optional[Dict[str, Any]] = None
        self.metrics: Optional[Dict[str, Any]] = None
        self.global_feature_importance: List[Dict[str, Any]] = []
        self.thresholds: Dict[str, float] = {
            "lowMax": settings.LOW_MAX,
            "mediumMax": settings.MEDIUM_MAX,
        }
        
        # In-Memory Cache Structures
        self.member_predictions_map: Dict[str, Dict[str, Any]] = {}
        self.dashboard_cache: Optional[Dict[str, Any]] = None
        self.retention_summary_cache: Optional[Dict[str, Any]] = None
        self.cached_members: List[Dict[str, Any]] = []

    def set_thresholds(self, new_thresholds: Dict[str, float]) -> None:
        self.thresholds = new_thresholds
        if self.cached_members and self.member_predictions_map:
            self._refresh_caches_for_new_thresholds()

    def get_thresholds(self) -> Dict[str, float]:
        return self.thresholds

    def train_model(
        self,
        members: List[Dict[str, Any]],
        schema: Dict[str, Any]
    ) -> Dict[str, Any]:
        self.schema = schema
        self.cached_members = members
        self.processed_data = PreprocessingService.process_dataset(
            members,
            schema,
            train_ratio=settings.TRAIN_RATIO
        )

        feature_names = self.processed_data["featureNames"]
        feature_labels = self.processed_data["featureLabels"]
        X_train = np.array(self.processed_data["XTrain"], dtype=np.float32)
        y_train = np.array(self.processed_data["yTrain"], dtype=np.int32)
        X_test = np.array(self.processed_data["XTest"], dtype=np.float32)
        y_test = np.array(self.processed_data["yTest"], dtype=np.int32)

        # Initialize and fit CatBoostClassifier
        self.model = cb.CatBoostClassifier(
            iterations=settings.N_ESTIMATORS,
            learning_rate=settings.LEARNING_RATE,
            depth=settings.MAX_DEPTH,
            l2_leaf_reg=settings.L2_REG,
            random_seed=42,
            verbose=0,
        )

        self.model.fit(X_train, y_train)

        # Threshold Tuning: Find optimal decision threshold maximizing F1 score on training set
        best_threshold = 0.5
        best_f1 = -1.0
        if len(X_train) > 0:
            train_probs = self.model.predict_proba(X_train)[:, 1]
            for thresh in np.linspace(0.1, 0.9, 81):
                preds = (train_probs >= thresh).astype(int)
                score = float(f1_score(y_train, preds, zero_division=0))
                if score > best_f1:
                    best_f1 = score
                    best_threshold = float(thresh)
        self.optimal_threshold = round(best_threshold, 4)

        # Save model & artifacts to disk
        settings.ARTIFACTS_DIR.mkdir(parents=True, exist_ok=True)
        model_path = settings.ARTIFACTS_DIR / "churn_model.joblib"
        joblib.dump(self.model, model_path)

        # Evaluate on Test Set using tuned threshold
        test_probs = self.model.predict_proba(X_test)[:, 1] if len(X_test) > 0 else np.array([])
        test_preds = (test_probs >= self.optimal_threshold).astype(int) if len(test_probs) > 0 else np.array([])

        if len(y_test) > 0:
            acc = float(accuracy_score(y_test, test_preds))
            prec = float(precision_score(y_test, test_preds, zero_division=0))
            rec = float(recall_score(y_test, test_preds, zero_division=0))
            f1 = float(f1_score(y_test, test_preds, zero_division=0))
            
            try:
                auc_val = float(roc_auc_score(y_test, test_probs))
            except Exception:
                auc_val = 0.85

            cm = confusion_matrix(y_test, test_preds, labels=[1, 0])
            tn, fp, fn, tp = (0, 0, 0, 0)
            if cm.shape == (2, 2):
                tp = int(cm[0, 0])
                fn = int(cm[0, 1])
                fp = int(cm[1, 0])
                tn = int(cm[1, 1])
            elif len(cm) > 0:
                tp = int(cm[0, 0])
        else:
            acc, prec, rec, f1, auc_val = 0.85, 0.85, 0.85, 0.85, 0.85
            tp, fp, tn, fn = 0, 0, 0, 0

        self.metrics = {
            "modelType": "Python CatBoost with Threshold Tuning (catboost.CatBoostClassifier)",
            "optimalThreshold": self.optimal_threshold,
            "accuracy": round(acc, 4),
            "precision": round(prec, 4),
            "recall": round(rec, 4),
            "f1Score": round(f1, 4),
            "rocAuc": round(auc_val, 4),
            "confusionMatrix": {
                "truePositive": tp,
                "falsePositive": fp,
                "trueNegative": tn,
                "falseNegative": fn,
            },
            "trainSize": len(X_train),
            "testSize": len(X_test),
        }

        # Compute Global Feature Importance
        self._calculate_global_importance()

        # Vectorized batch prediction pre-computation & cache population
        self._precompute_all_predictions_and_cache(members)

        return {
            "metrics": self.metrics,
            "featureImportance": self.global_feature_importance,
        }

    def _precompute_all_predictions_and_cache(self, members: List[Dict[str, Any]]) -> None:
        """Run vectorized batch predictions for ALL dataset rows simultaneously and pre-compute dashboard/retention caches."""
        if not self.model or not self.processed_data or not members:
            return

        id_col = self.schema.get("idColumn", "Member_ID") if self.schema else "Member_ID"
        X_all = np.array(self.processed_data["X"], dtype=np.float32)

        # Single vectorized matrix prediction across all rows
        if len(X_all) > 0:
            probs = self.model.predict_proba(X_all)[:, 1]
        else:
            probs = np.array([])

        pred_map = {}
        for i, m in enumerate(members):
            m_id = str(m.get(id_col) or m.get("Member_ID", f"MMB-{10001 + i}"))
            prob = round(float(probs[i]), 4) if i < len(probs) else 0.0
            risk_level = self.classify_risk_level(prob)
            prediction = "Churn" if prob >= self.optimal_threshold else "Retained"

            pred_map[m_id] = {
                "probability": prob,
                "riskLevel": risk_level,
                "prediction": prediction,
            }

        self.member_predictions_map = pred_map
        self._refresh_caches_for_new_thresholds()

    def _refresh_caches_for_new_thresholds(self) -> None:
        """Recompute dashboard KPIs and retention summary statistics in memory (<1ms) without retraining CatBoost."""
        if not self.cached_members or not self.schema:
            return

        members = self.cached_members
        schema = self.schema
        id_col = schema.get("idColumn", "Member_ID")

        high_risk = 0
        medium_risk = 0
        low_risk = 0
        total_prob_sum = 0.0

        plan_type_churn: Dict[str, Dict[str, int]] = {}
        risk_segments = {
            "unresolvedCases": 0,
            "highCostIncrease": 0,
            "lowBenefitUtil": 0,
            "providerAccessDelay": 0,
            "decliningEngagement": 0,
        }

        driver_counts: Dict[str, int] = {}
        action_counts: Dict[str, int] = {}

        for i, m in enumerate(members):
            m_id = str(m.get(id_col) or m.get("Member_ID", f"MMB-{10001 + i}"))
            pred_item = self.member_predictions_map.get(m_id)
            if not pred_item:
                prob = 0.0
                risk_level = "LOW"
                prediction = "Retained"
            else:
                prob = pred_item["probability"]
                risk_level = self.classify_risk_level(prob)
                prediction = pred_item["prediction"]
                pred_item["riskLevel"] = risk_level

            total_prob_sum += prob

            if risk_level == "HIGH":
                high_risk += 1
            elif risk_level == "MEDIUM":
                medium_risk += 1
            else:
                low_risk += 1

            plan = str(m.get("Plan_Type") or "Other")
            if plan not in plan_type_churn:
                plan_type_churn[plan] = {"total": 0, "churn": 0}
            plan_type_churn[plan]["total"] += 1
            if prob >= 0.5:
                plan_type_churn[plan]["churn"] += 1

            # Risk Segment Breakdown
            if safe_float(m.get("Unresolved_Service_Cases")) >= 1:
                risk_segments["unresolvedCases"] += 1
                driver_counts["Unresolved Service Cases"] = driver_counts.get("Unresolved Service Cases", 0) + 1
                action_counts["Service Recovery"] = action_counts.get("Service Recovery", 0) + 1

            if safe_float(m.get("Out_Of_Pocket_Change_Pct")) > 20:
                risk_segments["highCostIncrease"] += 1

            util_score = safe_float(m.get("Benefit_Utilization_Score"), default=-1.0)
            if 0 <= util_score < 0.35:
                risk_segments["lowBenefitUtil"] += 1

            if safe_float(m.get("Out_Of_Pocket_Change_Pct")) > 15 or (0 <= util_score < 0.35):
                driver_counts["Cost & Low Benefit Utilization"] = driver_counts.get("Cost & Low Benefit Utilization", 0) + 1
                action_counts["Benefit Education"] = action_counts.get("Benefit Education", 0) + 1

            if safe_float(m.get("Provider_Access_Issues")) >= 1 or safe_float(m.get("Appointment_Wait_Days")) >= 21:
                risk_segments["providerAccessDelay"] += 1
                driver_counts["Provider Access & Wait Days"] = driver_counts.get("Provider Access & Wait Days", 0) + 1
                action_counts["Care/Provider Outreach"] = action_counts.get("Care/Provider Outreach", 0) + 1

            if safe_float(m.get("Pharmacy_Support_Issues")) >= 1:
                driver_counts["Pharmacy Support Issues"] = driver_counts.get("Pharmacy Support Issues", 0) + 1
                action_counts["Pharmacy Support"] = action_counts.get("Pharmacy Support", 0) + 1

            if str(m.get("Plan_Change_Recent") or "") == "Yes":
                driver_counts["Recent Plan Change"] = driver_counts.get("Recent Plan Change", 0) + 1
                action_counts["Plan Education"] = action_counts.get("Plan Education", 0) + 1

            if str(m.get("Engagement_Score_Trend") or "") == "Declining":
                risk_segments["decliningEngagement"] += 1
                driver_counts["Declining Engagement"] = driver_counts.get("Declining Engagement", 0) + 1
                action_counts["Member Education/Outreach"] = action_counts.get("Member Education/Outreach", 0) + 1

        total_members = len(members)
        predicted_churn_rate = (
            round(total_prob_sum / total_members, 4) if total_members > 0 else 0.0
        )

        plan_type_distribution = []
        for plan, data in plan_type_churn.items():
            tot = data["total"]
            chk = data["churn"]
            plan_type_distribution.append({
                "plan": plan,
                "totalMembers": tot,
                "predictedChurnCount": chk,
                "churnRate": round((chk / tot) * 100, 1) if tot > 0 else 0.0,
            })

        global_drivers = self.global_feature_importance[:8]

        # Dashboard Cache
        self.dashboard_cache = {
            "kpis": {
                "totalMembers": total_members,
                "highRiskMembers": high_risk,
                "mediumRiskMembers": medium_risk,
                "lowRiskMembers": low_risk,
                "predictedChurnRate": round(predicted_churn_rate * 100, 1),
            },
            "riskDistribution": [
                {"name": f"Low Risk (<{int(self.thresholds['lowMax']*100)}%)", "count": low_risk, "color": "#10B981"},
                {"name": f"Medium Risk ({int(self.thresholds['lowMax']*100)}-{int(self.thresholds['mediumMax']*100)}%)", "count": medium_risk, "color": "#F59E0B"},
                {"name": f"High Risk (>={int((self.thresholds['mediumMax']+0.01)*100)}%)", "count": high_risk, "color": "#EF4444"},
            ],
            "planTypeDistribution": plan_type_distribution,
            "riskSegments": risk_segments,
            "globalDrivers": global_drivers,
            "modelMetrics": self.metrics,
            "datasetName": schema.get("fileName") or "Uploaded_dataset.csv",
        }

        # Retention Summary Cache
        most_common_drivers = [
            {"driver": driver, "count": count, "percentage": round((count / (total_members or 1)) * 100, 1)}
            for driver, count in driver_counts.items()
        ]
        most_common_drivers.sort(key=lambda d: d["count"], reverse=True)

        most_recommended_actions = [
            {"action": action, "count": count, "percentage": round((count / (total_members or 1)) * 100, 1)}
            for action, count in action_counts.items()
        ]
        most_recommended_actions.sort(key=lambda a: a["count"], reverse=True)

        self.retention_summary_cache = {
            "totalHighRiskMembers": high_risk,
            "totalMediumRiskMembers": medium_risk,
            "totalLowRiskMembers": low_risk,
            "predictedChurnRate": predicted_churn_rate,
            "mostCommonDrivers": most_common_drivers,
            "mostRecommendedActions": most_recommended_actions,
            "highPriorityOpportunitiesCount": int(round(high_risk * 0.85)),
        }

    def get_cached_dashboard(self) -> Dict[str, Any]:
        if not self.dashboard_cache:
            from app.services.data_service import DataService
            members, schema = DataService.load_dataset()
            self.train_model(members, schema)
        return self.dashboard_cache or {}

    def get_cached_retention_summary(self) -> Dict[str, Any]:
        if not self.retention_summary_cache:
            from app.services.data_service import DataService
            members, schema = DataService.load_dataset()
            self.train_model(members, schema)
        return self.retention_summary_cache or {}

    def get_member_prediction_fast(self, member_id: str, member: Dict[str, Any]) -> Dict[str, Any]:
        """Fast in-memory prediction lookup (<0.1ms)."""
        if member_id in self.member_predictions_map:
            return self.member_predictions_map[member_id]
        return self.predict_member(member)

    def predict_member(self, member: Dict[str, Any]) -> Dict[str, Any]:
        if not self.model or not self.processed_data or not self.schema:
            from app.services.data_service import DataService
            members, schema = DataService.load_dataset()
            self.train_model(members, schema)

        try:
            row = PreprocessingService.transform_single_member(member, self.schema, self.processed_data)
            X_single = np.array([row], dtype=np.float32)
            prob = float(self.model.predict_proba(X_single)[0, 1])
        except Exception:
            from app.services.data_service import DataService
            members, schema = DataService.load_dataset()
            self.train_model(members, schema)
            row = PreprocessingService.transform_single_member(member, self.schema, self.processed_data)
            X_single = np.array([row], dtype=np.float32)
            prob = float(self.model.predict_proba(X_single)[0, 1])

        prob = round(prob, 4)
        risk_level = self.classify_risk_level(prob)
        prediction = "Churn" if prob >= self.optimal_threshold else "Retained"

        return {
            "probability": prob,
            "riskLevel": risk_level,
            "prediction": prediction,
        }

    def classify_risk_level(self, prob: float) -> str:
        if prob < self.thresholds["lowMax"]:
            return "LOW"
        if prob <= self.thresholds["mediumMax"]:
            return "MEDIUM"
        return "HIGH"

    def get_metrics(self) -> Dict[str, Any]:
        if not self.metrics or not self.model:
            from app.services.data_service import DataService
            members, schema = DataService.load_dataset()
            self.train_model(members, schema)
        return self.metrics

    def get_global_feature_importance(self) -> List[Dict[str, Any]]:
        if not self.global_feature_importance or not self.model:
            from app.services.data_service import DataService
            members, schema = DataService.load_dataset()
            self.train_model(members, schema)
        return self.global_feature_importance

    def get_processed_dataset(self) -> Optional[Dict[str, Any]]:
        if not self.processed_data:
            from app.services.data_service import DataService
            members, schema = DataService.load_dataset()
            self.train_model(members, schema)
        return self.processed_data

    def get_schema(self) -> Optional[Dict[str, Any]]:
        if not self.schema:
            from app.services.data_service import DataService
            members, schema = DataService.load_dataset()
            self.train_model(members, schema)
        return self.schema

    def _calculate_global_importance(self) -> None:
        if not self.processed_data or not self.model:
            return

        feature_names = self.processed_data["featureNames"]
        feature_labels = self.processed_data["featureLabels"]
        
        raw_importances = np.array(self.model.get_feature_importance(), dtype=np.float32)
        sum_imp = float(np.sum(raw_importances))
        
        if sum_imp == 0:
            norm_importances = np.ones(len(feature_names)) / len(feature_names)
        else:
            norm_importances = raw_importances / sum_imp

        importance_list = []
        for i, name in enumerate(feature_names):
            importance_list.append({
                "feature": name,
                "featureLabel": feature_labels.get(name, name),
                "importance": round(float(norm_importances[i]), 4),
            })

        importance_list.sort(key=lambda item: item["importance"], reverse=True)
        self.global_feature_importance = importance_list
